### ISSUE TITLE
Squirrel.Windows - Reboot Proposal

### ISSUE THREAD

**Issue by shiftkey** [Squirrel/Squirrel.Windows#1470]
I wanted to provide an update to this situation that will hopefully address some concerns.

First off, many thanks to @anaisbetts for her hard work over the years on creating and maintaining Squirrel.Windows. As this thread shows, there’s a lot of love for the project and a lot of interest in keeping this going.

For those who don’t know me, I work at GitHub and have been around since the early days of Squirrel.Windows. I’m not an active contributor, but I’m a happy user of it due to its adoption in the Electron ecosystem and want to see it continue.

I had a chat with @robmen yesterday about how we might be able to help keep things going, and we agreed upon some short-term goals. These goals are rather modest, but they feel achievable over the coming weeks and months:

 - Review the outstanding open issues and get the issue tracker into a state that reflects the current situation
 - Hold a regular fortnightly meeting to discuss the project with contributors and users
 - Investigate whether there is some low-hanging fruit that can be worked on
 - Get the project to a place where we could potentially ship an update

We’re only two people with plenty of other things on our plate currently, so we’re looking for others to help with this work to build a more vibrant community around the project. 

**If this sounds like something you’d like to get involved with, please comment in this thread so we can gauge the interest.**

If you have any questions, feel free to comment here - or you can reach out privately to shiftkey@github.com if you don’t feel comfortable asking here.

**Comment by Inzanit** [comment]
I was saddened to see Squirrel.Windows be deprecated - just last month we managed to migrate most of our work apps to use Squirrel after having started using it late 2018. We've had a good experience. Upon the announcement me and a friend embarked on building a "spiritual-successor", if you will, that caters for our needs, both in "enterprise" and for commercial applications.

I'm glad to see Squirrel.Windows still has life left and would love to contribute where possible if you elaborate what help you're after.

**Comment by shiftkey** [comment]
> I'm glad to see Squirrel.Windows still has life left and would love to contribute where possible if you elaborate what help you're after.

The first thing that comes to mind is taking stock of the issue tracker. If anyone is able to spare some time, I'd love help with reviewing the open issues and seeing what we can close out. 

Things to look out for in open issues, and my suggested action:

 - Has the issue been resolved?
   - **Helpful:** Comment in the issue that you think this can be closed and mention me
 - Is the issue not clear? Is something missing that you think would help out?
    - **Helpful:** Ask a question that would help with understanding the problem and ensure that you `@-mention` the submitter so they're aware of it
 - Is this an old problem or a problem with an old version of the library?
    - **Helpful:** Ask the submitter if they are still able to reproduce the issue with the latest version of the library.
 - Is the issue asking a general question something the maintainers cannot really help with?
    - **Helpful:** Ask the submitter if they are fine to close out the issue, and ensure that you `@-mention` the submitter so they're aware of it

I can come through and help with sorting/closing issues, but having many eyes scanning over the open issues here would be a great help to get things moving...

**Comment by anaisbetts** [comment]
I'm glad to see that some folx found Squirrel so useful that they want to revive it :) I wanted to take a moment and share A few Guiding Principles that have served Squirrel.Windows well over the years.

* *Be as Dumb As Possible* - since the beginning, Squirrel.Windows has always tried to do literally the simplest, least clever thing that could possibly work. Simple things are great, because it means that people can build their own more complicated / things on top of it that only make sense for their scenario. Let Developers who Use Squirrel be the clever ones.  <br/><br/> Another extremely good reason to avoid complexity is, reliably installing and updating software on Windows is already walking a tightrope on a windy day while people throw tomatoes at you, complexity _will_ end up creating more failed updates and users getting left behind. Which leads me to....

* *Be extremely Conservative* - Squirrel.Windows routinely turns down features, moreso than other projects that I've maintained, because breaking the updater is Catastrophic. Slack broke the Mac updater shortly before moving to the Electron-based version of the app, and it took them _years_ to remediate this mistake, it sucks _really_ bad. Anything that touches the core path of updating the app should start off in your head with -5 votes. It sucks to turn people down, especially when they've already put in some work for their idea, but complexity adds up, and every feature you add will make it _harder_ to feel confident that you haven't broken the updater. 

* *Stick to The Principles and **Say No*** - If you buy #2 above, you're already sold on this, but for this project in particular, it is _so important_ to Say No. People want Squirrel.Windows to be a lot of different things, and Squirrel.Windows has stayed a reliable product because it is willing to tell people, "We just don't solve that problem. Sorry." Ideas like making Squirrel an NT Service to do updating would mean it would apply to more scenarios, but it would also make Squirrel much much more complicated and prone to leaving users stranded. 

* *Focus on the User Experience* - Another reason to Say No is because Squirrel.Windows has drawn a very strong Line In The Sand around user experience. The goal of an installer is, from the second the user double-clicks Setup.exe, we want to get the App in front of them, as fast as possible, and do _nothing_ else. When the user uninstalls? Our goal is to remove ourselves, as fast as possible - that's it! <br/><br/> And in terms of updating, our goal is to be like Chrome - you open it and it's the latest version. No prompts, no UAC dialogs, no Wizards or Progress Bars. Every time we show an update progress bar, we say to the user, "What you're doing isn't important. Look at Me instead". Any feature that tries to do this should be met with "That's Not What We Do Here"

Hope that helps explain some of the Zen of Squirrel.Windows. Excited to see what folx come up with in the future!

**Comment by peters** [comment]
Thanks for all the hard work you put into this project @anaisbetts 🎉

**Comment by frederikolsen88** [comment]
Yes, thank you for your work, @anaisbetts, and the best of luck on your journey!

And thank you for taking control, @shiftkey. What a relief 😄

**Comment by TonyValenti** [comment]
Hi All,
I just saw that aparently squirrel is deprecated?  If so, was there an announcement?

I'm very interested in keeping the project alive and well, especially with .NET Core 3.0 coming out soon, however, a lot of the way that code was written makes me feel fairly uncomfortable.  I'm more on board with a rewrite over maintenance.  Thoughts?

**Comment by shiftkey** [comment]
@TonyValenti 

>  especially with .NET Core 3.0 coming out soon, 

This is an interesting potential audience, and I'd love to hear more thoughts about this and how Squirrel.Windows might affect them.

> however, a lot of the way that code was written makes me feel fairly uncomfortable.

Can you elaborate on this? I'd love to understand these concerns and how we can mitigate them.

> I'm more on board with a rewrite over maintenance. Thoughts?

There is a lot of existing usage of Squirrel.Windows in the wild currently, and I'm very reluctant to contemplate rewriting things (and introducing breaking changes and potentially breaking these users) without more explicit goals and a plan to go about it.

**Comment by TonyValenti** [comment]
@shiftkey -
1.  With .NET core 3.0, desktop apps are now a supported scenario and I think that Squirrel.Windows makes a ton of sense for it.  If you deploy a desktop app, you're going to want a way to update it.  Also, I believe that the self-contained nature of .NET core is great for Squirrel.

2.  Here's a really good example.  Take a look at :
https://github.com/Squirrel/Squirrel.Windows/blob/master/src/Squirrel/Utility.cs

There is a ton of code in there that isn't documented and should really be broken out into multiple classes.  Additionally, you'll see things like Tuple returns and methods like this which make me shudder:
````
  public static Task ForEachAsync<T>(this IEnumerable<T> source, Func<T, Task> body, int degreeOfParallelism = 4)
        {
            return Task.WhenAll(
                from partition in Partitioner.Create(source).GetPartitions(degreeOfParallelism)
                select Task.Run(async () => {
                    using (partition)
                        while (partition.MoveNext())
                            await body(partition.Current);
                }));
        }
````

When I look at the code in general, it works but it needs some good love to start looking maintainable.

**Comment by robmen** [comment]
Yeah, we can definitely improve Squirrel.Windows but we don't need to start with a "rewrite". Small, iterative steps that bring along everyone to a better and better experience will work out well. A big bang rewrite would require significant backwards compatiblity verification and not allow us to see improvement along the way.

**Comment by aaronstine** [comment]
Hey guys, my company ([Identifi](http://Identifi.io)) recently adopted squirrel and wouldn't mind investing in its future. Our current product plan keeps our devs busy for the next few months. But after that, we could free up some folks to contribute.

**Comment by anaisbetts** [comment]
@TonyValenti Squirrel.Windows is an Ineffable Paragon of Software Engineering Elegance and I will hear nothing to the contrary!

**Comment by stefanolson** [comment]
I think the nice thing about squrirrel is how well it has worked. We've got a couple of thousand users using it on a daily basis and really we have quite minimal numbers of problems. Certainly better than anything else out there so well with continuing with I think.  I'm still running version 1.4.4. I did try and upgrade at one point but it started signing all the DLLs and taking ages to do the build, so I gave up and went back to 1.4.4. But we are now having a couple of problems with 1.4.4 so I will be downloading the code and submitting some changes if I can figure out the problem we are having.

**Comment by MeikTranel** [comment]
I would be more than willing to help migrate the project files to multitargeted projects where we can move Squirrel to .NET Core while keeping support for full framework. I must say... theres some complex stuff living in the project files. Maybe the current maintainers should make an effort to improve architecture documentation - the use cases are fairly well documented from what it looks like, but i don't see much information regarding how all the pieces fit together.
This is not a dig on the current maintainers... it's just me being interested to contribute but being a little overwhelmed with the architecture... and i assume we can all agree that more community participation is a factor that could really help this kind of project.

**Comment by Keboo** [comment]
I too would be willing to help out on keeping this project alive and moving forward.

**Comment by shiftkey** [comment]
Just wanted to give a shout out to @Thieum who has been investigating old issues and confirming things which can be closed out as being resolved! Thanks for your help!

**Comment by robmen** [comment]
@MeikTranel I am slowly working through the build process, simplifying as I can but primarily working towards a fully automated build and release process.

**Comment by TonyValenti** [comment]
@MeikTranel @robmen -
I'm working through a fairly substantial upgrade/cleanup of the squirrel.windows DLL.

**Comment by MeikTranel** [comment]
> @MeikTranel I am slowly working through the build process, simplifying as I can but primarily working towards a fully automated build and release process.

That'll definitely help me. Seeing how things move through the build ending up at the consumer is exactly the information i need regarding project file cleanup/migration.

I'm in a super tight spot in my education right now with finals coming up next week,  but i'm hoping to join the rest of the pack starting next thursday. There's a number of open source things i've got going, but currently on ice because of this reason.

@TonyValenti That's great, but i honestly feel like the more critical pillar of the codebase is the C++ stuff. There's some self-titled "janky" unzip code here, some "stub-setup" there... Maybe we should focus first on trimming things out which should not be Squirrel's primary domain. Without knowing to much of the architecture i'd say unzipping should be something we let .NET or any library do. It just feels to me that it shouldn't be my job to look at decompression code.

**Comment by TonyValenti** [comment]
@MeikTranel
I think that's a good idea.  I'm targeting the current .NET side of things because it aligns with a near-term need for our business.

Regarding the C++ side of things, I think it would be worth exploring to see if C++ can be removed from the stack and entirely replaced with .NET Core 3.0.

**Comment by robmen** [comment]
@TonyValenti arguably, Squirrel would benefit if everything was written in native code. That would remove the need for multi-megabyte external dependencies (.NET Framework is fortunately normally pre-installed so .NET Framework is not a distribution burden). .NET Core would add tens of megabytes and would dwarf many of the applications distributed via Squirrel.

Now installing .NET Core 3.0 runtime as a dependency (like .NET Framework today) will be an important feature in the not too distant future. But making Squirrel depend on it will have to take into account the whole of the Squirrel ecosystem.

**Comment by TT39** [comment]
I would be willing to help out. Squirrel is really useful software.

**Comment by MeikTranel** [comment]
@robmen i agree 100%. We should .NET Framework where it can be useful. .NET Standard for the referenced libraries and .NETCore/.NETFramework for the build time components. As for Native C++ i don't see a reason why we should forcibly remove it from the stack. As long as we don't try to solve problems that have been solved by the dependencies we already pull in the only real reasons to remove it would be consistency of the stack and the potential for xplat support; both of which are really weak arguments at the moment i reckon.

**Comment by anaisbetts** [comment]
> As for Native C++ i don't see a reason why we should forcibly remove it from the stack

You can't remove the C++ code because you can't guarantee that the version of .NET that you want is installed on the machine. The C++ code exists to install the .NET Framework.

> .NET Core would add tens of megabytes and would dwarf many of the applications distributed via Squirrel.

This is actually far worse, it's *hundreds* of megabytes at the moment. Mono's linker is much better at the moment, you can get Hello World down to 4MB, but you're still not beating the C++ bootstrapper.

**Comment by Thieum** [comment]
> .NET Core would add tens of megabytes and would dwarf many of the applications distributed via Squirrel.

Would [CoreRT](https://github.com/dotnet/corert), when ready, be an option to adress that issue?

**Comment by anaisbetts** [comment]
@Thieum afaik at the moment, CoreRT is similar or slightly bigger than the Mono Linker,

**Comment by MeikTranel** [comment]
> The C++ code exists to install the .NET Framework.

Does it install .NET Framework?

> This is actually far worse, it's hundreds of megabytes at the moment. Mono's linker is much better at the moment, you can get Hello World down to 4MB, but you're still not beating the C++ bootstrapper.

That's what i meant when i said we shouldn't try to forcibly remove it from the stack for seemingly no major reasons besides convenience - especially considering what you said about the downsides of the alternatives.

@Thieum 
> Would CoreRT, when ready, be an option to adress that issue?

We should be looking into that, but i don't expect it to be a magic switch as of right now. When .NET Core 3.0 is released, we should investigate this - the .NET team is doing lots and lots of work regarding this, AOT compiling and tree-shaking. I don't think its worth the effort right now.

**Comment by robmen** [comment]
CoreRT may be an answer but chances are it'll still be in the MB range. To be clear, there is a very real possibility to get Squirrel to add no more than 128K-512Kb overhead to the overall installation. Knowing that is a lower bound, I look critically at frameworks that add overhead measured in MB.

**Comment by anaisbetts** [comment]
@robmen Agree, though I would try to balance the concerns of disk space vs the ability for folx to contribute - it might be worth the extra MBs to get the maintenance help (and tbh, while network bandwidth is definitely still a Thing in some places, Internet speed is A Lot Faster than it used to be)

**Comment by robmen** [comment]
@anaisbetts Agreed. Knowing the "lower bound" allows us to measure the cost/benefit of each dependency.

**Comment by amaitland** [comment]
#### Issues

> Things to look out for in open issues, and my suggested action:

@shiftkey I'll propose a different action plan for the issues, setup [Probot Stale](https://github.com/probot/stale), close everything that's older than a set date, maybe 180 days? Set the initial bot message to something like `In the interest of bringing order to the repository old issues will be automatically closed unless confirmation the issue is still valid is received within x days`. Set the auto close to maybe `28` days or something. 

#### Branches

Cleanup all the  stale branches, deleting the `merged` ones should be a quick win https://github.com/Squirrel/Squirrel.Windows/branches/stale

It appears that `master` and `develop` are the two primary branches used, looking at https://github.com/Squirrel/Squirrel.Windows/blob/develop/docs/contributing/contributing.md and I don't see any clear guidance on the branch structure for the project, I think if the plan going forward is to have automated releases then a more clearly defined structure would be of benefit. Possibly just a `master` and `stable` branch system would be enough or my preference is branch per major release where they are prefixed with something like `release` so a generic `Branch Protection Rule` can be applied to them all. e.g. `release/1.9`

#### EditorConfig

I think adding a [.editorconfig](https://editorconfig.org/) file now before you get really going with new development is worthwhile, keep the codebase consistent 

#### Issue Templates

Add the standard bug report, feature request issues templates. in an effort to improve the quality of the opened issues.

#### Bots

There are a few bots that might be worth considering in the effort to maintain this repository

https://github.com/rsarky/helpr
https://github.com/probot/duplicate-issues
https://github.com/probot/no-response

**Comment by shiftkey** [comment]
@amaitland thanks for the suggestions:

Re: **issues**, I currently I don't want to bulk close issues because I still see some value in reviewing and interacting with them:

 - helps me to understand how others are using the project
 - helps to illustrate the sorts of problems users have encountered (whether fixed or outstanding)
 - helps identify gaps in documentation or opportunities to improve current docs
 - helps signal interesting areas to dive into about improving the project

Bulk closing things out to me feels like throwing away this information. We've already been able to close out over 150 issues and review over 110 more since I left that initial comment, which I think is some great progress (thanks to everyone else who has helped out).

As things settle down and we get into the harder stuff on the issue tracker, I'll revisit this and see what things might be worth automating.

Re: **branches** - I cleaned up most of those stale ones just now - there's a couple of interesting ones that I'll keep handy and poke at whether they might be useful after rebasing.

> I don't see any clear guidance on the branch structure for the project, I think if the plan going forward is to have automated releases then a more clearly defined structure would be of benefit

This is something @robmen and I have been chatting about. I've opened #1477 to track getting some docs together to explain what these changes mean in terms of PRs and releases.

Re: **editorconfig** - I've opened #1478 to track this, if someone would like to put one together

Re: **issue templates** - I've opened #1479 to remind myself to add this

**Comment by gojanpaolo** [comment]
Hey guys, just wanted to let you know that I'm interested in helping keep Squirrel.Windows alive. I primarily develop WPF apps using WiX and Squirrel.Windows for deployment solutions. I started sorting out the issues, will start submitting PRs later to fix some of it :) Cheers!

**Comment by Horb** [comment]
Thanks to @anaisbetts for all his work on Squirrel!

We use Squirrel for several of our applications and we intend to keep using and passing changes to upstream in the future.

**Comment by damieng** [comment]
I spent a fair bit of time looking into what a streamlined Squirrel next-gen might look like and agree with @anaisbetts that it really should do the dumbest amount possible.

For me that meant a Squirrel-built .exe should be able to:

1. Unpack the contents to a new folder
2. Add an entry into add/remove programs that will allow removal
3. Remove everything if launched from add/remove
4. Run a developer-specified command line to perform post-install steps
5. Run a developer-specified command line to perform pre-uninstall steps

I'd personally like to see app icons, auto-run, pining, file associations etc. entirely handed externally.  In Atom we moved most of this logic to atom itself where the app could offer users a rich ui and options for performing these steps without having to learn different tooling or languages.

I also thought that the unpacking should be pretty agnostic.  I looked into basically using a 7z self-extractor for that part but the idea that Squirrel next-gen wouldn't care what compression format you are using is appealing.  7z max-compression files were significantly smaller than zip and while not as small as the Squirrel delta system removed much of that complexity.

**Comment by TonyValenti** [comment]
@damieng -
I wanted to let you know I'm hard at work on a spiritual successor to Squirrel.  I'm starting with the easy things and moving to the harder ones (like self-installing).

**Comment by Miggleness** [comment]
Would moving Squirrel under .NET Foundation help?

**Comment by erwinwolff** [comment]
I would love to maintain Squirrel. My product needs to survive and it is by far the easiest installer I used to date. I would hate it to see it go down in flames.

**Comment by isatahiri** [comment]
Any status on Squirrel futur ? I just discovered and tested it. I really like it and want to know if we can rely on new project ?

**Comment by Lukkian** [comment]
So, @shiftkey how have you been?

I didn't mean to be rude, but could you give us any sign that you're still interested in maintain Squirrel.Windows?

I noticed that issues have dropped considerably in recent months, but some new ones have come up and it's been a long time since you commented here.

Before anyone judges me, I'm just a pragraming student and have some personal (non-commercial) projects that depend on Squirrel.Windows. Unfortunately (yet) I do not have the knowledge to contribute with code updates.

I'm really hoping Squirrel.Windows can move on.
Thanks for everything and God bless you.

**Comment by robmen** [comment]
I can't speak for @shiftkey but I still have great plans for Squirrel.

The timing of the project transition was far from ideal (it wasn't planned, obviously :smiley:). It was just before summer here. Summer means time spent on family vacations, traveling to kid's soccer tournaments, BBQs with friends, hiking and other fun stuff outside. It is not a great time to pick up an additional codebase and start making immediate progress. :smile: 

But the rain is coming in a couple months and then there will be more time spent at the keyboard. 

Before the summer started, I was able to massively improve the build process. Others have been doing a fantastic job triaging issues. So, we'll be in fantastic shape to roll forward.

**Comment by robmen** [comment]
@damieng I think there need to be "some batteries" included (auto-update, app icons, shortcuts, etc) in Squirrel. But I also believe that factoring Squirrel so you can use it without any of those batteries is a good goal as well. As I start working on v2.0 (mostly aligned with the notes @anaisbetts forwarded me before summer), I'll keep this design goal in mind.

**Comment by damieng** [comment]
@robmen the problem is everyone wants to do something a little different around that so you end up with a ton of customization options to deal with it which becomes a ton of work and a minefield.

Given your app comes with a full programming language it's so much easier to do exactly what you want there and just keep the installer down to install/uninstall/update.

**Comment by damieng** [comment]
This also allows the app to include add/remove from startup as a config option/setting, create icon on desktop/quicklaunch/start in case they removed it, add/remove exe from path....

**Comment by robmen** [comment]
@damieng *A lot* of people want the bare minimum (shortcut, app icon, ARP, auto-updates) taken care of for them. Squirrel needs to provide a solid implementation of those basics. At the same time, rather than expose lots of dials and knobs to customize Squirrel, the design will allow a developer to use Squirrel with none of the basics enabled. Then a developer can provide implementations for those basics as needed. That design provides most people what they want, avoids implementing tons of customizations in Squirrel, and does not block completely custom/advanced scenarios.

**Comment by damieng** [comment]
Hey, you're the one working on it! I will now bow out.

**Comment by Lukkian** [comment]
What a fantastic news @robmen, thank you so much for sharing it with us.
I will look forward to it. After all, the Squirrel we have today, is already a magnificent tool and I would be very happy just with bug fixes. New features alone is beyond what I expected, I am really happy.

**Comment by TonyValenti** [comment]
Hi All,
In case you're interested, I've been revamping a bunch of the code:
https://github.com/mediatedcommunications/badger

I don't have the installing/packaging done but I do have a lot of the framework for other things.

**Comment by mihaimyh** [comment]
Any plans to support .net-core?

**Comment by robmen** [comment]
@mihaimyh can you open a new issue with more details about the exacts support for .NET Core you'd like to see. Are you looking for it to install the redist instead of .NET Framework? If so (or if not :smile: ) say so in the new issue.

**Comment by Thieum** [comment]
@mihaimyh see also #1374 for existing discussion

**Comment by bzuillsmith** [comment]
Very cool to see the guy behind WiX Toolset working on this @robmen! Are you doing any kind of online pair or mob programming sessions or just currently architecting solo in your free time.

**Comment by robmen** [comment]
@bzuillsmith At this very moment, I'm working through the final steps of [a WiX Toolset release process for a security vulnerability][b]. This weekend I was at my son's soccer double-header on the other side of the mountains. So, right now (as mentioned above), I'm wrapping up a bunch of "stuff" from the summer.

When the rain rolls in a couple of months, I expect to have a bit more time to pick up the work I started in Squirrel before the summer and plot out a clear direction. Also as noted above, the timing of the transition was sudden and my goal is to make it clear the project has a future. But we simply were not going to make a lot of noise over the summer. :smiley: 

[b]: https://www.firegiant.com/blog/2019/9/12/wix-online-meeting-176-highlights/

**Comment by mihaimyh** [comment]
> @mihaimyh see also #1374 for existing discussion

Yes, I am already throwing my questions and suggestions there.

**Comment by BackGrounder84** [comment]
Hi, any update on this topic? I see no real progress for months.

**Comment by kylelaverty** [comment]
Hey, it has been several months since any update and over a month since the last time anyone followed up. I think there are a lot of people willing to help maintain this fantastic project and there are many tweaks that could be done without having the focus on v2.

I have not been involved with any public large projects before but I would be willing to help.

**Comment by varKeytrap** [comment]
Same here, this is a greate project and should be maintained!

**Comment by JuniorTrojilio** [comment]
I met the squirrel seeing apps like discord, postman and blitz use them for installation and updates, we are working on a new product in the company I work for and for this reason I am trying to think of an architecture that solves the main problems of installation, update and uninstallation , I am willing to contribute with the squirrel, I will try to know a little more about the project to be able to contribute and also use it

**Comment by rvoisin** [comment]
I've been using this across a large pile of apps for years now, and can make some time to keep the project going if need be.

**Comment by robmen** [comment]
First of all, I apologize for going dark. To say the last few months did not go as I expected would be an understatement. I anticipate my schedule to remain challenging for the next two months. It has been a long six months.

So you might be wondering what's holding up accepting the existing PRs. The first major undertaking I did was to get the continuous integration build working. That automates the verification that PRs will not break the build. The second major undertaking is to [develop automated integration tests to ensure PRs will not break updates by Squirrel](#1590). The absolutely worst case is if a bug is introduced that prevents Squirrel from updating.

Until such integration tests are in place, I am not comfortable integrating the bulk of the PRs nor doing extensive feature work.

That means, at this time, the best way to help Squirrel move forward is to contribute to the update integration tests for Squirrel (brainstorm scenarios, automate the production of example updates, help with infrastruc). I also very much appreciate all the work some have done to triage the issues so we have a clear idea of the work is ahead of us once updates are guaranteed.

Again, I'm sorry I've been absent for the last few months. I'm finally getting well and look forward to having the energy in a couple of months to make more progress on Squirrel.

**Comment by RudolfJan** [comment]
I recently heard about Squirrel and would love to use it. I invested some time, but just discovered .NET Core is not supported. For the installer feature, I cannot take that risk.  I do not really feel capable to contribute here, but I really hope some day Squirrel will support .Net Core. If necessary, I can work on the documentation, if there is any need.

**Comment by anaisbetts** [comment]
Squirrel is supported on .NET Core via the non-C# API, check out this Gist for an example of how it works:

https://gist.github.com/anaisbetts/9015f33a95c523a133bb624e3c77213d

**Comment by RudolfJan** [comment]
@anaisbetts Thank you for sharing this . I assume the general idea is the create a .Net Framework console wrapper application to handle squirrel updates? I'll see if I can get t working, should be possible ..

**Comment by JayBeavers** [comment]
While I get the dynamics here and the importance of never breaking the installer, both VSCode and GitHub Desktop are blocked from supporting Arm64 native builds by #1616 which is a minor dependency update from 2017.  Any chance someone with maintainer privs could just push through the PR for this low risk update from a dependency bugfix that's three years mature?

**Comment by peters** [comment]
Is anyone interested in an trying out our xplat update solution? It's been in production at our company since April 2019. It has a lot of features such as: built-in support for delta updates, release channels (test, staging, production) and automatic deployment using GitHub Actions. It uses Avalonia for xplat gui (offline and web installer support). Checkout https://github.com/fintermobilityas/snapx. The documentation is a bit scarce but a fully working demo can be found at https://github.com/fintermobilityas/snapx.demoapp. If you want to chat you can join our gitter community: https://gitter.im/fintermobilityas-snapx/community. NB! This is not a commercial product and will forever remain free. It's not a business area for us as a we are fin-tech company that wants to give back to the open source community :)

**Comment by bfourie** [comment]
Hi @anaisbetts , @robmen 
Our compagny want to use Squirrel.Windows to deploy and update our Software but with the following announcement we don't understand if this project is maintenaining.
I Hope you can enlighten us on that subject,
Regards.

**Comment by Sunbreak** [comment]
@anaisbetts is investigating Flutter. We'll see

**Comment by shiftkey** [comment]
I'm going to have to step down from any sort of maintainership work here for personal and professional reasons.

When I pitched this proposal I imagined having the bandwidth to support this important piece of community infrastructure, but since then my work responsibilities have taken me far from the Electron community, and my free time for anything open-source related is now limited. I've also been trying to look after my health these past few months, which is why I've been avoiding this hard discussion as well as other neglected projects on my plate. 

Best of luck with everything!

**Comment by IanKemp** [comment]
Since ClickOnce is back with .NET 5.0, what is the future of this project?

**Comment by robmen** [comment]
@IanKemp Nothing changes. Squirrel addresses the weaknesses in ClickOnce, not the fact that it continues to exist.

**Comment by caesay** [comment]
I'm now using this to deploy one of my net5 projects, so I've forked and updated everything to netstandard2/net5 and fixed a bunch of issues. I wouldn't consider it ready for general consumption but it's a good start. I also merged a couple PR's from here related to dotnet support. I will continue to improve it as my needs evolve, and I'm happy to review and accept PR's or discuss the shortcomings with anyone interested.

https://github.com/clowd/Clowd.Squirrel

At the time of writing, these are the changes that have been made, and there have been some feature regressions noted also.

Dependency / Framework Changes:
 - Nuget submodule removed and inlined. Updated to netstandard2.0
 - DeltaCompressionDotNet dependency removed and inlined. Updated to netstandard2.0
 - Removed ILRepack (replaced with PublishSingleFile)
 - All other nuget dependencies have been removed or updated to latest versions
 - Squirrel updated to netstandard2.0
 - Update/SyncReleases updated to net5.0 (w/PublishSingleFile)
 
Code Changes:
 - Removed lots of code from NugetCore that was not used to cut down on dependencies. 
 - Removed all usages of Assembly.Location as this does not work with PublishSingleFile
 - CreateShortcutForThisExe allows configurable shortcut locations 
 - Automatically get Uninstall icon from SquirrelAware executable instead of downloading from url
 - Update.exe icon can be optionally provided by dropping a 'ins-update.ico' file into project directory. Useful as rcedit currently can not replace icons in a PublishSingleFile executable so this must be done during Squirrel build.
 - SyncReleases supports downloading releases from 'web', 'github' or 'b2' (backblaze). It also now supports uploading new releases, and can be used to fully-automate a release deployment (eg. download last release, releasify build, upload new release)
 - Other misc bug fixes
 - build script has been replaced entirely, although produces a similar output that could be used with electron in theory
 - Setup.exe now bootstraps .net 5.0 desktop runtime instead of full fx.
 - Support for detecting dotnet core / net5 SquirrelAware binaries 
 
Known Regressions / Considerations before use:
 - Only x64 platform is supported by Setup/Update.
 - framework-version option is ignored. Currently, only .net 5.0 desktop runtime will be installed.
 - loadingGif option is ignored / there is no splash window.
 - I have not tested wix/msi support as I don't use it but I assume it will work.
 - rcedit can't be used to edit Update.exe icon
 
Possible improvements:
 - Explore ways to reduce binary size for NativeAOT or SelfContained as to remove runtime .net 5.0 dependency from Update.exe (will allow us a lot more flexibility when installing runtimes if the updater itself has no dependencies). Currently a SelfContained single file Update.exe binary is around 40mb but by removing dependencies carefully we should be able to get this comfortably under 10mb which might be good enough.
 - Add support for customisable framework-version and loadingGif again. (will likely need to do loadingGif with WinAPI to keep binary size down)
 - Fix and deploy as a nuget package: Clowd.Squirrel
 - Add .editorconfig and update all formatting in sln as there is now some inconsistency

**Comment by IanKemp** [comment]
> I'm now using this to deploy one of my net5 projects, so I've forked and updated everything to netstandard2/net5 and fixed a bunch of issues.

LIKE A BOSS.

**Comment by caesay** [comment]
Currently, [in my fork](https://github.com/clowd/Clowd.Squirrel), Update.exe ships as a .net5 `PublishSingleFile=true`, `SelfContained=false`. This produces a relatively small (2.6mb) single file binary that requires the .net5 runtime is installed during setup.

I have further explored the possibility of removing the runtime .net 5 dependency from Update.exe. This would be a huge win strategically because it would mean that non-dotnet apps (like electron) do not need to have .net core/fx bootstrapped at all during setup. For dotnet apps we would only need to bootstrap the framework the app depends on (or nothing, if the app we're deploying is self contained).

Using NativeAOT+UPX and no code changes we can achieve a final zero-dependency Update.exe binary size of 12mb. Not amazing, but if it saves us from downloading a 60mb dotnet runtime installer, worthwhile! Additional savings could be had by reducing code dependencies. The biggest issue will inevitably be that we use COM interop for creating shortcuts and updating tray icon settings. This is currently [not supported by NativeAOT](https://github.com/dotnet/runtimelab/issues/306). 

This leaves us with the following options then
- Rewrite Update.exe in C++ (an enormous endeavour, but would produce the best end results. I don't have time for this)
- Write a small native helper exe for shell/com interop that ships along side a NativeAOT compiled Update.exe
- Ship a standard SelfContained Update.exe (least work; but Update.exe binary size would be ~50mb)
- Bootstrap .net 5 in addition to `framework-version` (if different). non-dotnet apps will still need the net5 runtime installed during setup, and for dotnet apps we might end up needing to install two runtimes during setup. (eg. net5 to support Update.exe and netcore31 or net6 etc for the installed app)

My preference is the last option, it is easy to do, is unlikely to introduce issues, and produces the best experience if the target runtime(s) are already installed.

Interested to hear if anyone else has thoughts on the best approach or is willing to help with this? Once this is done, the fork will be close to feature parity, with the only real remaining drawback being the x64 platform requirement. Perhaps we can ship separate x86 and x64 versions of Updater.exe/Setup.exe to solve this if x86 support is needed.

**Comment by DavidZidar** [comment]
Wouldn't it make more sense to keep Update.exe as a .NET Framework app and make the API surface .NET Standard 2.0? That way Update.exe can stay small and the API can be used in both .NET Core and legacy .NET Framework applications. Microsoft will surely keep supporting .NET Framework for many years to come.

**Comment by caesay** [comment]
@DavidZidar The API surface (Squirrel itself) is already netstandard2.0 in my fork. Retargeting Update.exe the full framework (eg net48) instead of net5 is trivial, we'd need to use Fody or ILRepack instead of PublishSingleFile, and we end up in the same situation as 4th option above. We'd need to bootstrap net48 if it's not installed to support Update.exe, in addition to the framework/runtime required by the app we're installing... Although net48 is presumably shipped with all recent versions of Windows, so it might be safe enough to expect it to be installed.

**Comment by caesay** [comment]
I have released a [nuget package](https://www.nuget.org/packages/Clowd.Squirrel) and a [github release](https://github.com/clowd/Clowd.Squirrel/releases) containing all of the previously mentioned changes, as well as a few more, such as:

- Setup.exe now shows the splash image (instead of Update.exe). This means it can be shown quicker (no need to wait for zip to extract), and we don't need to carry around a heavy UI toolkit like WPF. (important for self contained option!)
- Setup.exe now defaults to installing no framework at all.. but can be used to install any modern framework (eg. `--targetFramework net5`) or even more than one framework (eg `--targetFramework net48,net5`).
- The `--selfContained` command line switch can be used to deploy a fully self-contained and dependency free Updater. This only adds +18mb uncompressed (+8mb compressed). Very reasonable!
- and more!

I have also updated the quickstart guide, and will look to start updating other docs soon.

See the whole release [here for more info](https://github.com/clowd/Clowd.Squirrel/releases/tag/2.1.7-pre)

**Comment by kant2002** [comment]
@caesay if you still want to pursue NativeAOT and what to solve your COM issue let me know. I think starting using [ComWrappers](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.comwrappers?view=net-5.0#remarks) instead of Built-in COM interop is a way to go in long-term. Built-in COM interop  would be disable by default in .NET 6 when publishing self-contained applications.
Using ComWrappers make code even more trim-friendly, and maybe shave couple bytes from application itself.

Example how it's done in .NET 6 is here https://github.com/dotnet/runtime/pull/54636 mostly that's pretty mundane changes.

**Comment by caesay** [comment]
@kant2002 I'm definitely interested in exploring options. Having a light weight, zero dependency updater is a key priority [in my fork](https://github.com/clowd/Clowd.Squirrel). I know in .NET 6 there are also several other improvements to trimming and `EnableCompressionInSingleFile` for example, that will allow us to reduce the size of a self contained Update.exe significantly. Would you be willing to submit a PR over there for com wrappers?

**Comment by Lazizbek97** [comment]
@anaisbetts  Thank you for creating such a great plugin(Squirrel.Windows). 
I have tried several times to use and integrate it into my app but didn't manage to do it. I would really appreciate it , if you share any example or doc on how to use it.

Or anyone who used it, please share your experience on usage.
